pvlib/ivtools.py

```diff
@@ -0,0 +1,349 @@
+
+    .. math::
+
+        I &\approx \frac{I_{L}}{1 + G_{p} \times R_{s}}
+        - \frac{G_{p} \times V}{1 + G_{p} \times R_{s}} \\
+          &= \beta_{0} + \beta_{1} \times V
+
+        3. The exponential portion of the IV curve is defined by
+            :math:`\beta_{0} + \beta_{1} \times V - I > ilim \times i_sc`.
+            Over this portion of the curve, :math:`exp((V+I*Rs)/(nNsVth)) >> 1`
+            so that
+
+    .. math::
+
+        \exp(\frac{V + I \times R_{s}}{nNsVth}) - 1 \approx
+        \exp(\frac{V + I \times R_{s}}{nNsVth})
+
+           Fit the exponential portion of the IV curve.
+
+    .. math::
+
+        \log(\beta_{0} - \beta_{1} \times V - I) &\approx 
+        \log(\frac{I_{0}}{1 + G_{p} \times R_{s}} + \frac{V}{nNsVth}
+        + \frac{I R_{s}}{nNsVth} \\
+        &= \beta_{2} + beta_{3} \times V + \beta_{4} \times I
+
+        4. Calculate values for ``IL, I0, Rs, Rsh,`` and ``nNsVth`` from the
+            regression coefficents beta0, beta1, beta3 and beta4.
+
+
+    References
+    ----------
+    [1] S.R. Wenham, M.A. Green, M.E. Watt, "Applied Photovoltaics" ISBN
+    0 86758 909 4
+    [2] C. B. Jones, C. W. Hansen, Single Diode Parameter Extraction from
+    In-Field Photovoltaic I-V Curves on a Single Board Computer, 46th IEEE
+    Photovoltaic Specialist Conference, Chicago, IL, 2019
+    """
+
+    # If not provided, extract v_oc, i_sc, v_mp and i_mp from the IV curve data
+    if v_oc is None:
+        v_oc = voltage[-1]
+    if i_sc is None:
+        i_sc = current[0]
+    if v_mp_i_mp is not None:
+        v_mp, i_mp = v_mp_i_mp
+    else:
+        v_mp, i_mp = _find_mp(voltage, current)
+
+    # Find beta0 and beta1 from linear portion of the IV curve
+    beta0, beta1 = _find_beta0_beta1(voltage, current, vlim, v_oc)
+
+    # Find beta3 and beta4 from the exponential portion of the IV curve
+    beta3, beta4 = _find_beta3_beta4(voltage, current, beta0, beta1, ilim,
+                                     i_sc)
+
+    # calculate single diode parameters from regression coefficients
+    return _calculate_sde_parameters(beta0, beta1, beta3, beta4, v_mp, i_mp,
+                                     v_oc)
+
+
+def _find_mp(voltage, current):
+    """
+    Finds voltage and current at maximum power point.
+
+    Parameters
+    ----------
+    voltage : ndarray
+        1D array containing voltage at each point on the IV curve, increasing
+        from 0 to v_oc inclusive, of `float` type [V]
+
+    current : ndarray
+        1D array containing current at each point on the IV curve, decreasing
+        from i_sc to 0 inclusive, of `float` type [A]
+
+    Returns
+    -------
+    v_mp, i_mp : tuple
+        voltage ``v_mp`` and current ``i_mp`` at the maximum power point [V],
+        [A]
+    """
+    p = voltage * current
+    idx = np.argmax(p)
+    return voltage[idx], current[idx]
+
+
+def _calc_I0(IL, I, V, Gp, Rs, beta3):
+    return (IL - I - Gp * V - Gp * Rs * I) / np.exp(beta3 * (V + Rs * I))
```

Comment:
Nit: This might be more immediately understood in terms of `nNsVth` instead of `beta3` = 1 / `nNsVth`.